### PR TITLE
DBZ-2522 Fixing index out of bounds for columns excluded from cdc tables

### PR DESCRIPTION
**Pull request by jgormley6**
https://issues.redhat.com/browse/DBZ-2522

Updated the data array in the IndicesMapping to match the source column width. This prevents index-out-of-bounds exceptions when you exclude columns in the middle of the table.

Example: A source column looking like [0,1,2,3,4], and the cdc table with [1,4]; the resulting data column for us should be [null, 1, null, null, 4] while the previous implementation would only have an array of size 2 and error out when placing the '4'.

### CODE REVIEW

**Comment by jgormley6** [comment]
I realize now this commit removed this change to fix tests.
https://github.com/debezium/debezium/pull/1758/commits/dd48172608b604648c586399550245e0cc75bf91
I think the issue is the test themselves, and not the array size. Digging into it

**Comment by jgormley6** [comment]
Updated the test to try a different failure condition around data type processing.
There were other failures, but they seemed unrelated to what I changed and might be flakey tests. Will let them rerun with the new commit.

**Comment by jgormley6** [comment]
It looks like only unrelated tests are intermittently failing now. If I need to increase the stability of these to advance this PR, just let me know and I can try to find time to dig into them.

File: debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java [hunk 1]
```diff
@@ -1290,6 +1290,85 @@ public void excludeColumnWhenCaptureInstanceExcludesColumns() throws Exception {
         stopConnector();
     }
 
+    @Test
+    @FixFor("DBZ-2522")
+    public void excludeColumnWhenCaptureInstanceExcludesColumnInMiddleOfTable() throws Exception {
+        connection.execute(
+                "CREATE TABLE blacklist_column_table_a (id int, amount integer, name varchar(30), primary key(id))");
```
Comment (gunnarmorling) [inline debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java:1297]: We're [moving away](https://debezium.io/blog/2020/09/03/debezium-1-3-beta1-released/) from the terms blacklist and whitelist. Could you please use another name, e.g. "table_with_excluded_column"?

**Review by gunnarmorling — COMMENTED** [review]
Thanks a lot, @jgormley6! This LGTM overall, one remark regarding naming inline.

In terms of the test failure, this looks unrelated indeed and you don't need to care about it. It's tracked via https://issues.redhat.com/browse/DBZ-2344 already // CC @Naros.

@jpechane any thoughts from you on this fix, as you're most familiar with this piece of code?

File: debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java (hunk 1 above)
Comment (jgormley6) [reply debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java:1297]: Yeah I can make that change! I was copying from an older branch and thought I changed all the occurrences, I'll make sure I didn't miss anymore

Comment (gunnarmorling) [reply debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java]: Thanks!

**Comment by gunnarmorling** [comment]
LGTM now. The failing test is a known flaky one (tracked as https://issues.redhat.com/browse/DBZ-2344). Thanks, @jgormley61!

**Comment by jpechane** [comment]
@jgormley6 LGTM thanks a lot. I'd like just ask you for two changes
1) Could you please not use column blacklist name but the new naming?
2) Could you make a copy of your current test and make it opposite - so instead of using column exclude list use column inlcude list?
The test and behaviour shloud be the same just different config.

**Comment by jgormley6** [comment]
I believe I changed all occurrences in my changes to use the new terminology. Should I be updating other existing test to use the new terms as well? 

I can add the test Sadly my internet is out this morning (using my phone for this), but I'll try to have that test up asap

**Comment by jpechane** [comment]
Yes please, `SqlServerConnectorConfig.COLUMN_BLACKLIST` etc. will be removed in future so it is better to keep to the new names in new code.

**Comment by jgormley6** [comment]
Oh sorry, I see, I missed the import for the config key. Didn't realize it. I'll fix that.

I quickly wrote the test, but its failing. 
It looks to me like COLUMN_INCLUDE_LIST config value is not being currently handled at all. 
I can try fixing it as part of this PR or we should create another story to implement it for SQL Server. Currently it looks like only postgres supports it, so would we also need to implement it for mysql (I'm not familiar with any connectors other than the SQL server one)?

**Comment by jgormley6** [comment]
It looks like the documentation for SQL server doesn't include column.whitelist or column.include.list either.

**Comment by jgormley6** [comment]
I updated to handle the include list for sql server the same way postgres handles it (include list and exclude list can't both be used and include list being the priority). Let me know if there's anything more here I should do for that.

**Comment by jpechane** [comment]
@jgormley6 you are way too fast :-). I intended to tell you it is not necessary to do that but obviously I am late to the game.

WHE the CI passes I'd like to ask you for two three sentences about this feature in SQL Server docs either as part of this PR or another one. Otherwise the code is good to go!

**Comment by jgormley6** [comment]
Updated the docs for the include list. I haven't worked with .adoc before so let me know if there's anything fishy about the formatting

**Comment by jpechane** [comment]
Rebased to the latest master and added contibutor to authors - https://github.com/debezium/debezium/pull/1812

**Comment by gunnarmorling** [comment]
Thanks a lot @jgormley6! Re flaky tests, yeah, that's a sore point atm. We're keeping track of them and work to get this sorted out, but it takes some time...

**Pull request closed without merging**